Source/JavaScriptCore/heap/MarkedBlock.cpp

```diff
@@ -502,6 +511,86 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
 }
 
+#if PLATFORM(COCOA)
+#define LOG_INVALID_HANDLE_DETAILS(s, ...) do { \
+    out.print("INVALID HANDLE: " s, __VA_ARGS__); \
+    WTF::setCrashLogMessage(out.toCString().data()); \
+} while (false)
+#else
+#define LOG_INVALID_HANDLE_DETAILS(s, ...) do { \
+    out.print("INVALID HANDLE: " s, __VA_ARGS__); \
+    dataLog(out.toCString().data()); \
+} while (false)
+#endif
+
+NO_RETURN_DUE_TO_CRASH NEVER_INLINE void MarkedBlock::dumpInfoAndCrashForInvalidHandle(AbstractLocker&, HeapCell* heapCell)
+{
+    JSCell* cell = bitwise_cast<JSCell*>(heapCell);
+    JSType cellType = cell->type();
+
+    StringPrintStream out;
+    LOG_INVALID_HANDLE_DETAILS("MarkedBlock = %p ; heapCell = %p ; type = %d\n", this, heapCell, cellType);
+
+    static_assert(!offsetOfHeader);
+    static_assert(!OBJECT_OFFSETOF(Header, m_handle));
+    size_t contiguousZeroCountAfterHandle = 0;
+    {
+        char* mem = WTF::bitwise_cast<char*>(&header());
+        for (; contiguousZeroCountAfterHandle < MarkedBlock::blockSize; ++contiguousZeroCountAfterHandle) {
+            if (*mem)
+                break;
+            ++mem;
+        }
+    }
+    LOG_INVALID_HANDLE_DETAILS("found %zd 0s at beginning of block\n", contiguousZeroCountAfterHandle);
+
+    bool isValidBlockVM = false;
+    bool isBlockInVM = false;
+    bool isBlockHandleInVM = false;
+    VM* blockVM = header().m_vm;
+    {
+        VMInspector::forEachVM([&](VM& vm) {
+            if (blockVM == &vm) {
+                isValidBlockVM = true;
+                LOG_INVALID_HANDLE_DETAILS("block VM %p is valid\n", &vm);
+                return IterationStatus::Done;
+            }
+            return IterationStatus::Continue;
+        });
+    }
+    if (isValidBlockVM) {
+        MarkedSpace& objectSpace = blockVM->heap.objectSpace();
+        isBlockInVM = objectSpace.blocks().set().contains(this);
+        isBlockHandleInVM = !!objectSpace.findMarkedBlockHandleDebug(this);
+        LOG_INVALID_HANDLE_DETAILS("block in our VM = %d, block handle in our VM = %d\n", isBlockInVM, isBlockHandleInVM);
+    }
+    if (!isBlockInVM && !isBlockHandleInVM) {
+        // worst case path
+        VMInspector::forEachVM([&](VM& vm) {
+            MarkedSpace& objectSpace = vm.heap.objectSpace();
+            isBlockInVM = objectSpace.blocks().set().contains(this);
+            isBlockHandleInVM = !!objectSpace.findMarkedBlockHandleDebug(this);
+            // Either of them is true indicates that the block belongs or used to belong to the VM.
+            if (isBlockInVM || isBlockHandleInVM) {
+                LOG_INVALID_HANDLE_DETAILS("block in another VM: %d, block in another VM: %d; other VM is %p\n", isBlockInVM, isBlockHandleInVM, &vm);
+                return IterationStatus::Done;
+            }
+            return IterationStatus::Continue;
+        });
+    }
+
+    uint64_t bitfield = 0xab00ab01ab020000;
```

Comment:
What's up with this? Is it just to add bits for yourself to sanity check? I don't think it's necessary but not the end of the world.

Reply:
Yes, just a sanity check to make sure we crashed at the end of the function.

Reply:
For future reference, I don't think it's necessary since` WTFCrashWithInfo` passes a bunch of `__LINE__, __FILE__, WTF_PRETTY_FUNCTION, __COUNTER__` to make it invalid to deduplicate. But not worth worrying about here.